client/components/isomorphic-id/index.js

```diff
@@ -0,0 +1,42 @@
+/** @ssr-ready **/
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+
+/**
+ * Internal dependencies
+ */
+import { incrementIsomorphicId } from 'state/ui/actions';
+import { getCurrentIsomorphicId } from 'state/ui/selectors';
+
+export default function addUniqueComponentId( Component ) {
+	class ComponentWithId extends React.Component {
+		constructor( props ) {
+			super( props );
+
+			this.componentId = this.props.currentIsomorphicId;
+			this.props.incrementIsomorphicId();
```

Comment:
Excellent point! The page I linked to has a good example on execution order. So even if a thunk resolved immediately we would also run into problems.
```javascript
function () {

  console.log('this is the start');

  setTimeout(function cb() {
    console.log('this is a msg from call back');
  });

  console.log('this is just a message');

  setTimeout(function cb1() {
    console.log('this is a msg from call back1');
  }, 0);

  console.log('this is the end');

})();

// "this is the start"
// "this is just a message"
// "this is the end"
// "this is a msg from call back"
// "this is a msg from call back1"
```